File: src/leetcode/python/medium/SumRootToLeafNumbers_129.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class SumRootToLeafNumbers_129(object):
    def sumNumbers(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
        stack, res = [(root, root.val)], 0
        while stack:
            node, val = stack.pop()
            if not node.left and not node.right:
                res += val
            if node.right:
                stack.append((node.right, val * 10 + node.right.val))
            if node.left:
                stack.append((node.left, val * 10 + node.left.val))
        return res

File: src/leetcode/java/SumRootToLeafNumbers_129.java
package leetcode.java;

public class SumRootToLeafNumbers_129 {
  private int sum;

  public int sumNumbers(TreeNode root) {
    if (root == null) return 0;
    
    if (root.left == null && root.right == null) return root.val;
    dfs(root.left, root.val * 10);
    dfs(root.right, root.val * 10);
    return sum;
  }

  private void dfs(TreeNode node, int subSum) {
    if (node == null) return;
    subSum += node.val;
    if (node.left == null && node.right == null) {
      sum += subSum;
      return;
    }
    dfs(node.left, subSum * 10);
    dfs(node.right, subSum * 10);
  }

  class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
  }
}
